feat: add aws cdk example

Yengas opened the pull request:
Adding an example to run `openapi-backend` with AWS API Gateway using the REST API Spec and a single Lambda. Using AWS CDK for infra.

anttiviljami commented:
Hi @Yengas ! Thanks for the contribution!

TBH I feel like your sample is a far too opinionated. I'd like a more minimal CDK example, hopefully with the `NodeJSFunction` construct only.

Your package.json includes a lot of good but opinionated dev tooling / pracrtices I'd prefer to leave out. For example:

- projen
- @aws-lambda-powertools
- eslint config

Yengas commented:
Hey @anttiviljami, 

Thanks for the feedback. Looking back at it, the MR seems bloated to me from the context of `openapi-backend` as well.

I can disable `eslint`, remove `@aws-lambda-powertools` and eject `projen`. 

I do think [SpecRestApi](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-apigateway.SpecRestApi.html) provides a good example regarding how much you can get for free by using OpenAPI spec, `openapi-backend` and AWS. WDYT?

anttiviljami commented:
Hmm, I'm a bit torn because SpecRestApi uses API Gateway v1, which has a lot of redundant overlapping functionality with openapi-backend, such as routing and validation.

Do you see some clear benefits in recommending people to use AWS REST APIs over the more lightweight v2 HTTP APIs with openapi-backend?

Yengas commented:
Hey @anttiviljami,

Not an expert on the differences, however just researched a little bit now. REST API seems more costly, however has more features that you can use to off-load work to the API GW(Routing, Validation, WAF), instead of handling things in your Lambdas or with additional resources.

You are right on the fact that it has overlapping features with `openapi-backend` though. Keeping this simple with a V2 HTTP API and proxing all requests to a single lambda started to make more sense to me as well. Will refactor to that, in the following week 👍

Yengas commented:
@anttiviljami I've updated the example per our discussion and the overall MR got simplified a lot! It now includes:

1. Single `NodeJsFunction` that runs `openapi-backend` just like `examples/aws-sam`. I took liberty with modifying the code a little bit and put in some opinionated defaults to the Lambda configuration. Hope those are not deal breakers for you!
2. `HttpApi` that only handles CORS and proxies all non `OPTIONS` calls to `NodeJsFunction`. No validation, routing etc. are done at this level.
3.  Plain AWS CDK. Ejected Projen and removed all non-critical dependencies.

Sorry for the late update and I wish you a happy holiday!

anttiviljami commented on examples/aws-cdk/package.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "license": "Apache-2.0",
```
Please use the correct license from the main package

```suggestion
  "license": "MIT",
```

anttiviljami commented on examples/aws-cdk/src/lambdas/api-entrypoint.lambda.ts:
```diff
@@ -0,0 +1,71 @@
+    {
+      method: event.requestContext.http.method,
+      path: event.requestContext.http.path,
+      // @ts-ignore
+      query: event.queryStringParameters,
+      body: event.body,
+      // @ts-ignore
+      headers: event.headers,
```
I feel a bit uneasy about the `@ts-ignore` comments in the minimal example. Let's find a way to avoid this.

anttiviljami commented on examples/aws-cdk/src/main.ts:
```diff
@@ -0,0 +1,80 @@
+    });
+
+    // let's create a separate proxy integration instead of using `$default` integration of `HttpApi`
+    // this way; `OPTIONS` requests will be handled by `HttpApi` instead of invoking your Lambda
```
Very nice approach 👌

anttiviljami commented on examples/aws-cdk/tsconfig.json:
```diff
@@ -0,0 +1,35 @@
+    "strictNullChecks": true,
+    "strictPropertyInitialization": true,
+    "stripInternal": true,
+    "target": "ES2019"
```
I think we can use ES2022 for Node 18:

https://github.com/tsconfig/bases/blob/main/bases/node18.json#L8

```suggestion
    "target": "ES2022"
```

anttiviljami reviewed (changes requested):
Looks very good. Just have a couple of suggestions, but will probably end up merging anyways a bit later 👍

anttiviljami commented:
@Yengas could you also set up the standard test for CI? 🙏

https://github.com/anttiviljami/openapi-backend/blob/master/.github/workflows/ci.yml#L15-L25

Yengas commented:
@anttiviljami fixed your comments by applying the suggestions locally. Added tests according to how other examples implement it. I had to downgrade to Node 16.x for the AWS SAM to work though, seems like it does not have 18.x images.

Tests are working locally atm, tried to add to Github actions the same way as AWS SAM.

anttiviljami commented on examples/aws-cdk/index.test.ts:
```diff
@@ -0,0 +1,42 @@
+
+jest.setTimeout(90000);
+
+describe('aws sam example', () => {
```
```suggestion
describe('aws cdk example', () => {
```

anttiviljami commented:
Wow great work @Yengas! Looks like the test failed due to an ancient version of Node.js. Mind rebasing your branch and pushing again to re-run the tests?

Yengas commented:
@anttiviljami applied the suggestion and rebased 👍
